internal/app/handlers/shortener_handler.go

```diff
@@ -99,21 +117,109 @@ func (s *ShortenerHandler) ShortenerJSONHandler(res http.ResponseWriter, req *ht
 
 	res.Header().Set("content-type", "application/json")
 
+	var responseJSON JSONResponse
+	var headerStatus int
+	var shortURL string
 	shortURLData, err := s.service.DecodeURL(shortenerRequest.URL)
+	isURLExists := false
 	if err != nil {
-		http.Error(res, "error decode url", http.StatusBadRequest)
-		return
+		var pgErr *pgconn.PgError
+		if errors.As(err, &pgErr) && pgErr.Code == storage.CodeErrorDuplicateKey {
+			isURLExists = true
+		} else {
+			http.Error(res, "error decode url", http.StatusBadRequest)
+			return
+		}
 	}
 
-	responseJSON := JSONResponse{
-		Result: fmt.Sprintf("%s/%s", config.AppConfig.BaseShortURL, shortURLData.ShortURL),
+	if isURLExists {
+		modelURL, err := s.service.Storage.FindByURL(shortenerRequest.URL)
+		if err != nil {
+			http.Error(res, "error find model", http.StatusBadRequest)
+			return
+		}
+		headerStatus = http.StatusConflict
+		shortURL = modelURL.ShortURL
+	} else {
+		headerStatus = http.StatusCreated
+		shortURL = shortURLData.ShortURL
+	}
+	responseJSON = JSONResponse{
+		Result: fmt.Sprintf("%s/%s", config.AppConfig.BaseShortURL, shortURL),
 	}
 	responseString, err := json.Marshal(responseJSON)
 	if err != nil {
 		http.Error(res, "error json marshal response", http.StatusInternalServerError)
 		return
 	}
 
+	res.WriteHeader(headerStatus)
+
+	_, err = res.Write(responseString)
+	if err != nil {
+		http.Error(res, "error write data", http.StatusBadRequest)
+		return
+	}
+}
+
+type BatchRequest struct {
+	CorrelationID string `json:"correlation_id"`
+	OriginalURL   string `json:"original_url"`
+}
+
+type BatchResponse struct {
+	CorrelationID string `json:"correlation_id"`
+	ShortURL      string `json:"short_url"`
+}
+
+// ShortenerBatch обработка списка адресов
+func (s *ShortenerHandler) ShortenerBatch(res http.ResponseWriter, req *http.Request) {
+
+	bodyValue, err := io.ReadAll(req.Body)
+	if err != nil {
+		http.Error(res, "error read bodyValue", http.StatusBadRequest)
+		return
+	}
+
+	defer req.Body.Close()
+
+	var requestItems []BatchRequest
+	if err = json.Unmarshal(bodyValue, &requestItems); err != nil {
+		http.Error(res, "error unmarshal json request", http.StatusBadRequest)
+		return
+	}
+	urls := make([]string, 0)
+	for _, requestItem := range requestItems {
+		if !s.regexURL.MatchString(requestItem.OriginalURL) {
+			continue
+		}
+		urls = append(urls, requestItem.OriginalURL)
+	}
+	modelURLs, err := s.service.DecodeURLs(urls)
+	if err != nil {
+		http.Error(res, "error decode urls", http.StatusBadRequest)
+		return
+	}
+	res.Header().Set("content-type", "application/json")
+
+	responseItems := make([]BatchResponse, 0)
```

Comment:
Мы же знаем длину массива, следовательно можем сразу указать capacity.
```suggestion
	responseItems := make([]BatchResponse, 0, len(responseItems))
```